keps/sig-api-machinery/2835-prioritized-leader-election/README.md

```diff
@@ -0,0 +1,324 @@
+
+
+### Goals
+
+* Provide a simple mechanism to allow users to opt into prioritized leader election.
+
+### Non-Goals
+
+* User prioritized leader election in Kubernete's maintained controllers
+* Require prioritized leader election to be used by all users of leader election
+
+## Proposal
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a controller implementor, I have multiple versions of my controller running concurrently in the
+cluster; I want the newest one to become the leader.
+
+#### Story 2
+
+As a controller implementor, I have multiple replicas with different resources; I want the largest
+replica to be the leader.
+
+#### Story 3
+
+As a controller implementor of a MCS controller, I want the geographically closest region to be the
+leader, to avoid something like a US controller acting on a cluster in
+Asia, when I have a controller replica already running in Asia.
+
+### Notes/Constraints/Caveats (Optional)
+
+N/A
+### Risks and Mitigations
+
+The largest risk here is that this is a new field. This means that users will need to account for older API versions,
+or controllers that do not understand prioritized leader election.
+
+This is mitigated by documenting that users need to take this into account.
+
+
+## Design Details
+
+A new configuration field will be provided to the leader election configuration:
+
+```go
+type KeyComparisonFunc func(existingKey string) bool
+
+type LeaderElectionConfig struct {
+...
+
+	// KeyComparison defines a function to compare the existing leader's key to our own.
+	// If the function returns true, indicating our key has high precedence, we will take over
+	// leadership even if their is another un-expired leader.
+	//
+	// This can be used to implemented a prioritized leader election. For example, if multiple
+	// versions of the same application run simultaneously, we can ensure the newest version
+	// will become the leader.
+	//
+	// It is the responsibility of the caller to ensure that all KeyComparison functions are
+	// logically consistent between all clients participating in the leader election to avoid multiple
+	// clients claiming to have high precedence and constantly pre-empting the existing leader.
+	//
+	// KeyComparison functions should ensure they handle an empty existingKey, as "key" is not a required field.
+	//
+	// Warning: when a lock is stolen (from KeyComparison returning true), the old leader may not
+	// immediately be notified they have lost the leader election.
+	KeyComparison KeyComparisonFunc
+}
+```
+
+The API of an election lock will also be changed. Note this is the same change in two places - one defines
+the `Lease` API and the other is a mirror of the API used for all leader election types.
+
+```go
+type LeaderElectionRecord struct {
+...
+	// HolderKey is the Key of the lease owner. This may be empty if a key is not set.
+	HolderKey            string      `json:"holderKey"`
+}
+
+type LeaseSpec struct {
```

Comment:
In the API, I think we want to indicate what the weighting function is and an integer weight.  That allows for different versions of a controller-manager to change its weighting function and for other leaders to be able to make an intelligent decision about what to do when they don't understand that weighting function.

For instance, when the lease  is expired, old clients that do not understand the weighting function could still take the lease and simply indicate a weight of 0.

Reply:
Ah, I didn't understand this in the call but I get it now. I am not sure an integer will support all cases.

Some examples

---

Newest version wins (string):
example key: `v1.8.1`
example key comparison function: `return semver(myVersion).NewerThan(semver(existingKey)`

Newest version wins (int):
example key: `8` (attempt to look at only the minor version
example key comparison function: `return myVersion > existingKey`

The issue with int here is we are trying to encode a SemVer into a single integer. There are ways around this, but not a good one. For example, if we later need to distinguish beyond minor version we cannot.

---

Selected controller wins (our primary use case

string:
example key: `my-controller-identifier`
example key comparison function: `return existingKey != selectedIdentifier && myIdentifier == selectedIdentifier` (where selectedIdentifier is determined out-of-band)

int: not really possible

---

Select geographically closest controller

string:
example key: `us-west`
example key comparison function: `return myKey is closer to TargetRegion than existingKey`

int: not really possible unless we encode each region as an integer


---